### ISSUE TITLE
Around 2 % overestimation for density under NPT simulation by OpenMM compared with other softwares

### ISSUE THREAD

**Issue by z-gong** [openmm/openmm#2593]
Hello, I performed NPT simulation at 300 K and 1 bar with Langevin integrator and MonteCarloBarostat on a simple LJ-126 liquid (Martini coarse-grained water model, vdW interactions are truncated at 1.2 nm without dispersion correction) with the script attached. The density I get is 1.039 g/mL, which is 1.7 % larger than the value produced by GROMACS and LAMMPS (1.022 g/mL). For GROMACS, I'm using Langevin integrator and Paranello-Rahman barostat. For LAMMPS, I'm using Nose-Hoover thermostat and barostat.

I also tried the new Nose-Hoover integrator in the master branch, and the overestimation is the same. I compared the energy calculated on exactly the same coordinates, OpenMM, GROMACS and LAMMPS give identical potential energy. It seems energy calculation and thermostat are correct, so the problem can only come from barostat.

I also tested SPC/E water at 300 K and 1 bar, the density I get is 1.017 g/mL, 2 % larger than literature value (0.998 g/mL), whereas Lammps and GROMACS give the correct result.

Is there any mistake in my script or is this difference expected between MonteCarloBarostat and virial based barostats? 2 % overestimation of density may sound small, but it means there's a huge difference in pressure.

```python
import sys, random
from simtk import openmm as mm, unit
from simtk.openmm import app

system = mm.System()
system.setDefaultPeriodicBoxVectors([4.8,0,0],[0,4.8,0],[0,0,4.8])
for i in range(1000):
    system.addParticle(72.06)

nbforce = mm.NonbondedForce()
nbforce.setNonbondedMethod(mm.NonbondedForce.CutoffPeriodic)
nbforce.setCutoffDistance(1.2)
nbforce.setUseDispersionCorrection(False)
for i in range(1000):
    nbforce.addParticle(0, 0.47, 5.0)
system.addForce(nbforce)

barostat = mm.MonteCarloBarostat(1 * unit.bar, 300)
system.addForce(barostat)

top = app.Topology()
chain = top.addChain()
for i in range(1000):
    residue = top.addResidue('WAT', chain)
    top.addAtom('W'+str(i), None, residue)

integrator = mm.LangevinIntegrator(300, 1, 0.01)  # the suggested step size is 0.03 ps
sim = app.Simulation(top, system, integrator)
sim.reporters.append(app.StateDataReporter('martini.log', 1000, step=True,
                     potentialEnergy=True, temperature=True, density=True,
                     speed=True, separator='\t'))

positions = [[i*0.48+random.random()/10, j*0.48+random.random()/10, k*0.48+random.random()/10]
    for i in range(10) for j in range(10) for k in range(10)] 
sim.context.setPositions(positions)
sim.context.setVelocitiesToTemperature(300)
print(sim.context.getState(getEnergy=True).getPotentialEnergy().value_in_unit(unit.kilocalorie_per_mole))
sim.minimizeEnergy(100)
print(sim.context.getState(getEnergy=True).getPotentialEnergy().value_in_unit(unit.kilocalorie_per_mole))
sim.step(200000)
```

**Comment by jchodera** [comment]
Wow, 2% is a very large density anomaly.

Do you have your complete script for recording the densities and averaging them, out of curiosity? Would be useful if we had a complete example to reproduce what you are seeing.

**Comment by peastman** [comment]
Have you tried plotting density as a function of time?  That will show, for example, how quickly it's equilibrating, and whether the density is consistently too high or is experiencing occasional spikes that raise the average.

**Comment by z-gong** [comment]
I've uploaded all the input files for OpenMM, GROMACS and LAMMPS. I also tested an LJ-12-4 liquid (SDK coarse-grained water), the density is overestimated by over 4 %. 
https://send.firefox.com/download/35959e17d72ba185/#qyjFKeOMQtaMjlrHHKCGvg

For the density, I just take the last 100 values from the log file and average them. The time evolution of density is quite normal in all cases. And from the trajectory, I see no crystallization.

I noticed that density tolerance for water box in the test for MonteCarloBarostat is 2 %, so I was wondering it might be normal
https://github.com/openmm/openmm/blob/ef02e48dbc09f2a6f8e8ce2b22010887d2585fe2/tests/TestMonteCarloBarostat.h#L259-L268

**Comment by z-gong** [comment]
Here is the time evolution of density. The systems reach equilibration soon. The density produced by OpenMM is systematically higher.

![image](https://user-images.githubusercontent.com/379616/76448902-66a7b800-63cb-11ea-856d-b9fde323cd48.png)

**Comment by peastman** [comment]
I can reproduce your results.  The density consistently comes out in the range of ~1.04.  I tried varying lots of settings, none of which made a difference:

- Increasing the barostat frequency.
- Using mixed precision.
- Using LangevinMiddleIntegrator.
- Reducing the step size.

I agree this is most likely related to the barostat.  Does either Gromacs or LAMMPS offer a MonteCarloBarostat?  It would be nice if we could test that variable directly.

**Comment by z-gong** [comment]
There's some clue. If I turn on the long-range dispersion correction, then Lammps and OpenMM give similar density. Without dispersion correction, OpenMM overestimates density. It seems to have something to do with the energy evaluation during Monte Carlo scaling. Still, it doesn't explain why the density of SPC/E is overestimated.

As far as I know, other than OpenMM, the only software that supports MC barostat is Amber. I've never used Amber before. The control script of Amber is really confusing. It will take some time for me to compare the results of Amber.

**Comment by z-gong** [comment]
Here are the results of three software on an LJ-12-6 fluid with epsilon = 0.8 kcal/mol and sigma = 0.47 nm (Martini water will freeze with dispersion correction, so I decreased the epsilon)

  | thermostat | barostat | cut = 1.2 nm with dispersion correction | cut = 1.2nm w/o dispersion correction
-- | -- | -- | -- | --
LAMMPS | Nose-Hoover | Nose-Hoover | 0.951 | 0.884
Amber | Langevin | Monte Carlo | 0.951 | 0.885
OpenMM | Langevin | Monte Carlo | 0.946 | 0.915

My previous comment is wrong. There is still a small discrepancy when dispersion correction is applied.

**Comment by peastman** [comment]
It's interesting, though, that the difference gets much smaller when you add the dispersion correction: from 0.03 without down to 0.005 with.  That suggests this really is something with the energetics.  I wonder if there could be some other difference in there.  For example, is Amber or LAMMPS using a switching function?  Are they all calculating the dispersion correction in the same way?  I searched through the Amber manual, but couldn't find a formula or reference for it.  The version in OpenMM is described at http://docs.openmm.org/latest/userguide/theory.html#lennard-jones-interaction.

**Comment by giacomofiorin** [comment]
This is quite interesting. @peastman you mentioned having tested mixed precision, do you have results for double precision as well?

**Comment by peastman** [comment]
Good thought.  I tried and it didn't make a difference.

**Comment by peastman** [comment]
As an experiment, I tried enabling a switching function to smoothly turn off the interaction between 1.1 and 1.2 nm:

```python
nbforce.setUseSwitchingFunction(True)
nbforce.setSwitchingDistance(1.1)
```

When I did that, the density increased from 0.946 to 0.951, which just happens to be exactly the value from LAMMPS and Amber.  Coincidence?  Quite possibly.  But definitely worth checking on.

**Comment by z-gong** [comment]
Both LAMMPS and Amber are using continuum model correction for dispersion energy and pressure.
![image](https://user-images.githubusercontent.com/379616/76562150-05ecae00-64a5-11ea-903e-54c52143f27b.png)
I'm not using switching function in any case. Actually Amber does not support LJ switching.

**Comment by peastman** [comment]
That's different from the correction in OpenMM.  We use the one from https://pubs.acs.org/doi/10.1021/jp0735987.

**Comment by ajsilveira** [comment]
Lammps uses the same formula described at  http://docs.openmm.org/latest/userguide/theory.html#lennard-jones-interaction.
Here is the code https://github.com/lammps/lammps/blob/8f3680039b65b157334948c72b5b7da180c55253/src/pair_lj_cut.cpp#L562
and then the division by the volume is done here https://github.com/lammps/lammps/blob/fa7085be077d9271f8400c27c63613cf1a652144/src/compute_pe.cpp#L99
Would it make sense to check the behavior when using a larger cut-off (2 nm or so) without dispersion corrections?

**Comment by peastman** [comment]
Sure, it's worth a try.  At any rate, it will be another data point to use in figuring out where the difference is coming from.

**Comment by z-gong** [comment]
For 2.0 nm cutoff without dispersion correction, LAMMPS gives 0.936 g/mL, OpenMM gives 0.942 g/mL. Indeed, the difference is smaller.

Could it be possible that LJ particles diffuse too fast while neighbour list is not updated sufficiently often?

**Comment by ajsilveira** [comment]
In lammps, are you getting any warning message regarding “dangerous” neighbor builds?

**Comment by peastman** [comment]
In OpenMM, the neighbor list is rebuilt whenever a particle has moved by half the padding distance, so interactions are never missed.

It's interesting that without the dispersion correction, OpenMM gives a higher density than LAMMPS (implying stronger attraction between particles?), but with the dispersion correction it gives a lower density than LAMMPS.  That seems like it might be significant.

**Comment by ajsilveira** [comment]
@z-gong in lammps you could add the following line to ensure that the neighbor list is rebuilt every step:
neigh_modify every 1 delay 0 check no

**Comment by z-gong** [comment]
@ajsilveira LAMMPS didn't complain about neighbour build.
The simulation result is the same after setting `neigh_modify every 1 delay 0 check no`

**Comment by ajsilveira** [comment]
I think it would be helpful to see a comparison of the potential energies in NVT simulations. In lammps you can obtain the contribution from the dispersion correction (which is gonna be a constant) by specifying 
```thermo_style custom evdwl etail```
evdwl contains etail so to plot them separately you just need to do evdwl-etail
To compute the dispersion term in openmm, you will have to compute it "manually" from the specified volume. Let me know if I can help.

**Comment by z-gong** [comment]
I messed up the Amber simulations. The results in the previous table for Amber is actually with Berensen barostat. (I was confused by the ntt, ntp and barostat parameters) The following table is the correct results.

Apparently, OpenMM matches the results of Amber exactly. So where does the discrepancy between MC and virial based barostat come from, especially without dispersion correction? Is it possible that when MC barostat scales the atom positions, the interactions between atoms nearby cutoff are not handled appropriately? e.g. two particles are separated by 1.19 nm, after MC scaling, it becomes 1.21 nm. Without dispersion correction, there will be a sudden increase in the potential energy, which prevents the new positions to be accepted, thus the density being overestimated.

  | thermostat | barostat | cut = 1.2 nm with dispersion   correction | cut = 1.2nm w/o dispersion   correction
-- | -- | -- | -- | --
LAMMPS | Nose-Hoover | Nose-Hoover | 0.951 | 0.884
Amber | Langevin | Berendsen | 0.951 | 0.885
Amber | Langevin | Monte Carlo | 0.947 | 0.914
OpenMM | Langevin | Monte Carlo | 0.946 | 0.915

The Amber script is attched to make sure the results are correct this time
```
NoDispCorr
 &cntrl
  imin=0,
  ntx=1,
  irest=0,
  nstlim=400000,
  dt=0.010,
  ntpr=1000,
  ntwx=1000,
  cut=12.0,
  ntt=3,
  temp0=300.0,
  ig=-1,
  gamma_ln=1.0,
  ntp=1,
  barostat=2,
  mcbarint=25,
  pres0=1.0
 /
 &ewald
  vdwmeth=0,
 /
```

**Comment by peastman** [comment]
So it's really a difference between barostats rather than a difference between programs.  I have a hypothesis about what's happening.  It's just a hypothesis, so we need to test it.  If it's correct, the Monte Carlo results are accurate and the other barostats are inaccurate.

Continuous barostats are based on integrating a quantity which is assumed to be continuous.  Whenever you try to numerically integrate a quantity that has discontinuities, you get errors because the answer isn't actually the integral (unless you explicitly include the delta functions at the discontinuities).  In this case, we're using a sharp cutoff on the interaction, so every time a particle crosses that discontinuity the barostat will accumulate a little error.  Monte Carlo barostats aren't affected by discontinuities though.  They just compare two discrete values and don't assume it's smooth in between.

The way to test this is to use a switching function to eliminate the discontinuities.  Once we do that, the difference between continuous and Monte Carlo barostats should disappear.  I posted results above for OpenMM with a switching function.  You said Amber doesn't support that option.  What about LAMMPS?

**Comment by z-gong** [comment]
Amber supports force switch, OpenMM supports potential switch. Here is the comparison with GROMACS. 

After removing the discontinuity, the results are quite close. Then I suppose we should just avoid plain cut-off under any circumstances.

  | Barostat | Cutoff 1.2 nm (no switch, no disp corr) | Potential switch (1.0 - 1.2 nm) | Force switch (1.0 - 1.2 nm)
-- | -- | -- | -- | --
GROMACS | Parinello-Rahman | 0.883 | 0.913 | 0.861
Amber | Monte Carlo | 0.914 |   | 0.858
OpenMM | Monte Carlo | 0.915 | 0.913 |  

I'll check again the results for SPC/E water, to make sure there's really discrepancy.

**Comment by peastman** [comment]
> Then I suppose we should just avoid plain cut-off under any circumstances.

That's actually a very complicated question.  You can look at it from several different perspectives, and they suggest different answers.

Any kind of cutoff will create artifacts.  A sharp cutoff creates artifacts of one sort.  A switching function creates artifacts of a different (usually but not always less severe) sort.  From this perspective, the best option is to use dispersion PME, since that eliminates the need for cutoffs.  With time I expect that to become increasing common.

On the other hand, you can look at it this way.  All MD force fields are poor approximations to the real physics.  We try to compensate for that by parametrization, fitting the force field to match experimental data or quantum calculations.  From this perspective, the "best" settings to use are whatever was used during parametrization.  If the force field was fit while using a sharp cutoff, that will probably give a better match to experiment than dispersion PME, even if the latter is technically "more accurate".

Another way to look at it is to recognize that every simulation has many sources of error.  Cutoffs are just one of them.  With many of those sources, you can adjust settings to trade off accuracy versus speed (such as by changing the cutoff distance).  You need to decide what is the maximum level of error that is acceptable for your application.  Then you try to choose settings to keep the total error from all sources below that threshold, while still making it as fast as possible.

Yet another perspective on the question: for a particular application, some types of error are more important than others.  For example, a sharp cutoff leads to poor energy conservation.  In some applications that's not a big problem.  In others it's a very big problem.  So you need to identify what types of error are and aren't acceptable in your particular application and choose settings accordingly.

To summarize, it's really complicated!  If there were a single set of settings that were always best, we'd just use them.  But there isn't, and that's why there are so many options.

**Comment by z-gong** [comment]
I agree that it's usually best to just follow the original settings of how the model was parameterized. But in this specific case about truncation, it's important to ensure that the energy equals the integral of force. Since the MD is propagated by force instead of energy, if the energy gets truncated sharply, the generated configurations will not follow Boltzman distribution. If we care about properties like cohesive energy, then we are actually simulating the system with (shifted) potential function but analyzing the results with another potential function.

About the density overestimation of SPC/E model, it turns out to have nothing to do with the barostat. It's because the CharmmPsfFile parser doesn't set up angle constraints correctly. I'll submit a pull request to fix it. Such a coincidence.

**Comment by peastman** [comment]
> It's because the CharmmPsfFile parser doesn't set up angle constraints correctly. I'll submit a pull request to fix it.

Awesome, thanks!
